src/Functions/FunctionTokens.h

```diff
@@ -106,9 +107,13 @@ class FunctionTokens : public IFunction
             const ColumnString::Chars & src_chars = col_str->getChars();
             const ColumnString::Offsets & src_offsets = col_str->getOffsets();
 
-            res_offsets.reserve(src_offsets.size());
+            res_offsets.resize_exact(src_offsets.size());
             res_strings_offsets.reserve(src_offsets.size() * 5);    /// Constant 5 - at random.
-            res_strings_chars.reserve(src_chars.size());
+            std::optional<size_t> res_chars_reserve_size = generator.getResultReserveSize();
+            if (res_chars_reserve_size.has_value())
```

Comment:
Sorry, I find the new resizing logic too implicit (not obvious enough) for future readers to understand. I also have trouble understanding its sense.

Assume we are in `splitByChar` (where `getResultReserveSize()` returns != `std::nullopt):
- In that case, `SplitByCharImpl::getResultReserveSize()` returns `src_chars.size()` (as initialized earlier by `init()`). L. 114 will do `res_strings_chars.reserve_exact(src_chars.size())`.

Assume we are not in `splitByChar`: (*)
- In that case, l. 116 will do `res_strings_chars.reserve(src_chars.size())`.

The latter case may allocate more memory than minimally needed. But performance is much more impacted by the number of allocations done than the memory size of individual allocations (assuming the allocated memory sizes are not huge which is not the expected case here).

In l. 137, the difference between the cases will be `resize_assume_reserved` vs. `resize`. The former only changes the end pointer, the latter potentially does a true reserve operation. In my view, the previous logic (*) (<-- see above) will have made sure that the "potential reserve operation" never takes place because the output structure was already sufficiently sized.

I also wonder why the fancy resize code is only implemented for "splitByString" but not for "splitByChar", "splitByRegexp", "splitByWhitespace", and "splitByNonalpha"?

But perhaps it makes no performance difference at all, it would be cool if you could measure if it actually helps performance.

Reply:
1. `reserve_exact` in `splitByChar` vs `reserve` in (*)
Yes. It makes no performance difference, except the former may consume less peak memory. 

2. `resize_assume_reserved` in `splitByChar` vs `resize` in (*)` 
I thought the former has better performance under the assumption that `resize_assume_reserved` has less instructions than `splitByChar` and it has no branch, which may bring some benefits in hot loops. I'll measure it later. 

3. > I also wonder why the fancy resize code is only implemented for "splitByString" but not for "splitByChar", "splitByRegexp", "splitByWhitespace", and "splitByNonalpha"?
Actually `getResultReserveSize` is only implemented in `splitByChar`. It is waiting to be done for other functions, maybe in another PR, if we are sure that it will bring some performance benefits.

Reply:
@rschu1ze I'm pretty sure that `resize_assume_reserved` behaves faster than `resize` in tight loop even if `resize` doesn't actually reserve. 

Query: `with 'Many years later as he faced the firing squad, Colonel Aureliano Buendia was to remember that distant afternoon when his father took him to discover ice.' as s select splitByChar(' ', materialize(s)) as w from numbers(1000000) format Null;`

Use `resize_assume_reserved` in loop.
```
localhost:9001, queries: 20, QPS: 2.235, RPS: 2234840.530, MiB/s: 17.050, result RPS: 0.000, result MiB/s: 0.000.
```

Use `resize` in loop
```
localhost:9001, queries: 20, QPS: 1.901, RPS: 1901012.773, MiB/s: 14.504, result RPS: 0.000, result MiB/s: 0.000.
```

Reply:
According to above test results, current implementation has the best performance.